### PR TITLE
Frontend/request modal

### PR DESCRIPTION
**Pull request by awilli70**
This PR is mostly frontend changes, though it contains minor tweaks to how timezones work.
Requests will require a bit of an overhaul on the backend, so this serves to get the ball rolling
Ideally the changes to requests will not take too long, and submission can be implemented.

Question: Should how do we want to handle multiple requests to the same timeslot? Do we want to give requests a 1/2 week expiry, and block others from making requests during that same time?

Modal upon clicking Request:
Screen Shot 2020-06-03 at 10 28 12 PM

Searchable Dropdown for availabilities:
Screen Shot 2020-06-03 at 10 29 25 PM

Regular Dropdown for topics:
Screen Shot 2020-06-03 at 10 29 34 PM

Submit Request is only enabled when availability and topic are selected:
Screen Shot 2020-06-03 at 10 29 43 PM

Text area and corresponding states in action:
Screen Shot 2020-06-03 at 10 30 38 PM

### CODE REVIEW

File: client/src/components/AlumniDirectory.js [hunk 1]
```diff
@@ -197,7 +197,7 @@ export default class AlumniDirectory extends Component {
             } else {
                 if (this.state.allText[i].match(searchPattern) !== null) {
                     numResults += 1
-                    display.push(this.constructProfile(this.state.entries[i]), i)
+                    display.push(this.constructProfile(this.state.entries[i], i))
```
Comment (awilli70) [inline client/src/components/AlumniDirectory.js:200]: visual bug now fixed

File: client/src/components/RequestModal.js [hunk 2]
```diff
@@ -147,42 +147,39 @@ export default class RequestModal extends Component {
...
-                    days.push({
-                        key: day.day.toLowerCase(),
-                        text: day.day,
-                        value: day.day
-                    })
-                }
-            }
-        }
-        this.setState({dayOptions: days})
+    async createOptions(availabilities) {
```
Comment (awilli70) [inline client/src/components/RequestModal.js:165]: Here is the bulk of the playing nice explained - the offset returned by JS getTimeOffset makes no sense whatsoever with our date model, and results in weird behavior - basically, I needed to make sure that 

A) Timezones were being stored relative to UTC (they weren't, but only because of offset weirdness) - FIXED
B) Because user timezones were being stored improperly, this caused a lot of otherwise functional code to break - this resulted in weirdness where after saving a preference and refreshing, the time shown would be completely different for alumni, and availabilities would be bizarre in the request modal
C) This meant weirdness when passing student timezone as a prop, which would basically double the offset in the wrong direction.

In short - it now works
I do have a question, which I will bring up during standup -
The applyTimezone in all/:timezone doesn't seem to be doing anything - this is fine, but performance wise it doesn't make sense to fix the timezone for all alumni when only a few will be requested in one session

The payload can be changed to remove the math, but to do so we will have to pass the student/alumni as a prop from App.js - that's fine, but it would also be a good first opportunity to use the store

File: server/routes/alumni.js [hunk 3]
```diff
@@ -41,7 +41,7 @@ router.post('/', async (req, res, next) => {
                 //requests: [{type: Schema.Types.ObjectId, ref: 'requestSchema'}]
                 //posts: [{type: Schema.Types.ObjectId, ref: 'postSchema'}]
                 availabilities: availabilities,
-                timeZone: timeZone,
+                timeZone: -timeZone,
```
Comment (awilli70) [inline server/routes/alumni.js:44]: Ironically this was all that was needed to get most things to play nice

File: server/routes/students.js [hunk 4]
```diff
@@ -28,7 +28,7 @@ router.post('/', async (req, res, next) => {
                 grade: grade,
                 // requests: [{type: Schema.Types.ObjectId, ref: 'requestSchema'}]
                 // issuesLiked: [{type: Schema.Types.ObjectId, ref: 'issueSchema'}]
-                timeZone: timeZone,
+                timeZone: -timeZone,
```
Comment (awilli70) [inline server/routes/students.js:31]: same here

File: server/routes/requests.js [hunk 5]
```diff
@@ -11,6 +12,22 @@ router.get('/', async (req, res, next) => {
     })
 })
 
+router.patch('/applyRequesterTimezone', async (req, res, next) => {
```
Comment (awilli70) [inline server/routes/requests.js:15]: This isn't necessary if we change how /alumni/all/:timezone works BUT it probably makes more sense to do it this way

File: client/src/components/AlumniDirectory.js [hunk 6]
```diff
@@ -49,18 +50,36 @@ props:
...
+            numResults: 0,
+            filter: 'all',
+            requestModalOpen: false,
+            alumniDetails: null,
+        }
+        this.openRequestModal = this.openRequestModal.bind(this)
+        this.closeRequestModal = this.closeRequestModal.bind(this)
+    }
+
+    closeRequestModal() {
```
Comment (reshadbinharun) [inline client/src/components/AlumniDirectory.js:74]: nit: can be changed to use one handler `toggleRequestModalOpen` that uses `setState({requestModalOpen: !this.state.requestModalOpen})`

File: client/src/components/AlumniDirectory.js [hunk 7]
```diff
@@ -183,6 +221,11 @@ export default class AlumniDirectory extends Component {
         }
         this.setState({ [name]: value })
     }
+    handleRequestButton(e) {
+        this.setState({alumniDetails: this.state.entries[e.currentTarget.dataset.id]})
+        console.log(this.state.alumniDetails)
```
Comment (reshadbinharun) [inline client/src/components/AlumniDirectory.js:226]: Remove `console.log`

File: client/src/components/AlumniDirectory.js [hunk 8]
```diff
@@ -267,6 +310,14 @@ export default class AlumniDirectory extends Component {
                 
                 {searchRow}
                 {resultsRow}
+                {this.state.requestModalOpen &&
```
Comment (reshadbinharun) [inline client/src/components/AlumniDirectory.js:313]: Do we need the `this.state.requestModalOpen && `?

File: client/src/components/RequestModal.js [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+    async createAvailabilityOptions(availabilities) {
+        console.log(availabilities)
+        /* 
+         * Offset is in minutes, annoyingly, and in the opposite direction 
+         * that you'd expect. This means that if you are in UTC -4 (EST), 
+         * it returns 240 (not -240 to reflect being 4 hours behind)
+         * That's why there's a tiny bit of math in the payload, to conform
+         * with our date model (which makes more sense)
+         */
+        let timeOffset = await new Date().getTimezoneOffset()
```
Comment (reshadbinharun) [inline client/src/components/RequestModal.js:74]: Do we need the `await` here?

File: client/src/components/RequestModal.js [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+
+    async componentWillMount() {
+        await this.setState({alumni: this.props.alumni})
+        this.createAvailabilityOptions(this.state.alumni.availabilities)
+        this.createTopicOptions(this.state.alumni.topics)
+    }
+
+    async createAvailabilityOptions(availabilities) {
+        console.log(availabilities)
+        /*
```
Comment (reshadbinharun) [inline client/src/components/RequestModal.js:67]: Thank you for documenting :)

File: client/src/components/RequestModal.js [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+        console.log(availabilities)
+        /* 
+         * Offset is in minutes, annoyingly, and in the opposite direction 
+         * that you'd expect. This means that if you are in UTC -4 (EST), 
+         * it returns 240 (not -240 to reflect being 4 hours behind)
+         * That's why there's a tiny bit of math in the payload, to conform
+         * with our date model (which makes more sense)
+         */
+        let timeOffset = await new Date().getTimezoneOffset()
+        console.log(timeOffset)
```
Comment (reshadbinharun) [inline client/src/components/RequestModal.js:75]: Remove `console.logs`

File: server/routes/alumni.js (hunk 3 above)
Comment (reshadbinharun) [inline server/routes/alumni.js:44]: So does this need to be `timeZone * -1`? (Not sure)

File: server/routes/requests.js [hunk 12]
```diff
@@ -11,6 +12,22 @@ router.get('/', async (req, res, next) => {
...
+    try {
+        const agnosticAvailabilities = req.body.availabilities;
+        const timezone = req.body.offset
+        const availabilities = await timezoneHelpers.applyTimezone(agnosticAvailabilities, timezone)
+        res.status(200).send({
+            availabilities: availabilities
+        })
+    } catch (e) {
+        console.log(e)
```
Comment (reshadbinharun) [inline server/routes/requests.js:24]: nit: May be we can include information about which route failed in `console.error`

**Comment by reshadbinharun** [comment]
Looks solid! A few comments.
As for your questions on request setup, I think we should:
a) Remove availability when an alumnus confirms a call for a given time slot. This might mean that when we fetch availabilities for an alumnus on modal open, we filter out a slot that corresponds to a `confirmed request by the same alumnus`. We can push this off to version 2 if this is too much work, let me know your thoughts
b) I haven't played with the idea of expiry on records in mongo but that does seem like a very powerful tool to keep the system clean! I think we out to set an expiry date of a week since our availabilities are weekly times.

File: client/src/components/AlumniDirectory.js (hunk 8 above)
Comment (awilli70) [reply client/src/components/AlumniDirectory.js:313]: Yes - otherwise it will try to render without an alumni prop

File: server/routes/alumni.js (hunk 3 above)
Comment (awilli70) [reply server/routes/alumni.js:44]: nope!
